hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3083,6 +3083,52 @@ def ndim(self):
         """
         return self._type.ndim
 
+    @property
+    def T(self):
+        """Reverse the dimensions of this ndarray. For an n-dimensional array `a`,
+        a[i_0, ..., i_n-1, i_n] = a.T[i_n, i_n-1, ..., i_0].
+        Same as `self.transpose()`.
+
+        See also :func:`.transpose`.
+
+        Returns
+        -------
+        :class:`.NDArrayExpression`.
+        """
+        return self.transpose()
+
+    @typecheck_method(axes=nullable(tupleof(int)))
+    def transpose(self, axes=None):
+        """Permute the dimensions of this ndarray according to the ordering of `axes`. Axis `j` in the `i`th index of
+        `axes` maps the `j`th dimension of the ndarray to the `i`th dimension of the output ndarray.
+
+        Notes
+        -----
+        Does nothing on ndarrays of dimensionality 0 or 1.
+
+        Returns
+        -------
+        :class:`.NDArrayExpression`.
+        """
+        if axes is None:
+            axes = list(reversed(range(self.ndim)))
+        else:
+            if len(axes) != self.ndim:
+                raise ValueError(f'Must specify a complete permutation of the dimensions.'
+                                 f'Expected {self.ndim} axes, got {len(axes)}')
+
+            if len(set(axes)) != len(axes):
+                raise ValueError(f'Axes cannot contain duplicates: {axes}')
+
+            for axis in axes:
+                if not 0 <= axis < self.ndim:
+                    raise ValueError(f'Invalid axis: {axis}')
+
+        if self.ndim < 2:
+            return self
+
+        return type(self)(NDArrayReindex(self._ir, axes), self._type)
```

Comment:
this is wrong -- should use construct_expr and pass indices/aggs through.

Reply:
Ah right that's what I needed. Should I never then explicitly construct expressions with the class constructors?

Reply:
I don't think so. If I could have made those private in Python, I would have. Using construct_expr everywhere ensures there's only one place that contains the logical mapping from hail type => typed expression